File: algorithms/46. Permutations/solution.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def dfs(i):
            if i >= len(nums):
                result.append(nums[:])
            for j in xrange(i, len(nums)):
                nums[i], nums[j] = nums[j], nums[i]
                dfs(i + 1)
                nums[i], nums[j] = nums[j], nums[i]
        result = []
        dfs(0)
        return result

# 25 / 25 test cases passed.
# Status: Accepted
# Runtime: 72 ms
# beats 75.27 %

File: algorithms/46. Permutations/solution.cc
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        dfs(nums, 0, result);
        return result;
    }

    void dfs(vector<int>& nums, int start, vector<vector<int>> & result) {
        if ( start >= nums.size() ) {
            result.push_back(nums);
        }
        for ( int i = start; i < nums.size(); ++i ) {
            swap(nums[i], nums[start]);
            dfs(nums, start + 1, result);
            swap(nums[i], nums[start]);
        }
    }
};
